### ISSUE TITLE
Multiple Output Streams - C++ Error

### ISSUE THREAD

**Issue by Xeonzinc** [geekuillaume/node-audioworklet#4]
Hi,

I am trying to output to two audio devices at once on the same machine. Individually they work correctly, however when both are setup to play at once (with duplicate AudioServer objects etc....) only one plays and I get the following error:

![image](https://user-images.githubusercontent.com/8190861/90913746-9de7af80-e3d4-11ea-87e9-2facf9e54b84.png)

It seems to run the processFrames function of both devices about 2/3 times, before seemingly randomly picking one that will play, while the other process frame is not called again (I assume because of the above error).

 Is there an appropriate approach which could avoid this?

**Comment by geekuillaume** [comment]
I've changed and cleaned a lot of things in this lib. Can you try again and tell me if you still have the same problem?

**Comment by Xeonzinc** [comment]
Hi, I've just tested with v5.0.0 and still get an identical error.

**Comment by geekuillaume** [comment]
Can you send me the code you're using to get this error?

**Comment by Xeonzinc** [comment]
Apologies I've been away for a few days. My basic code has stopped working the the latest version, I'm not sure if the the pushAudioChunk needs to be called in processFrame now? The white noise example also doesn't seem to run right now so I'm struggling to get it working with your latest release (5.0.2) right now.

I've attached my (currently non working) code below for reference anyway, as I don't think it changes the issue with multiple outputs.

```
const AudioServer = require('audioworklet').AudioServer;
AudioServer.setDebugLog(false);
const audioServer = new AudioServer();
var fs = require('fs');
var lame = require("@suldashi/lame");

var assigned_device_lookup = new Object();


var buffSamp = new Buffer.allocUnsafe(2);
buffSamp.writeInt16LE.call(buffSamp, 100);

assigned_device_lookup.audio_1 = {'name':"Speakers (3- USB PnP Sound Device)"};
assigned_device_lookup.audio_2 = {'name':"Speakers (USB PnP Sound Device)"};

const logDevice = (device) => {
	for (var key of Object.keys(assigned_device_lookup)) { 
		if (assigned_device_lookup[key].name == device.name) {
		  	  
			//Setup the object and mp3 decoding streams
			assigned_device_lookup[key].device = device;
			assigned_device_lookup[key].decoder = new lame.Decoder();
			assigned_device_lookup[key].key = key;		 
			var file = "C:\\Users\\emina\\OneDrive\\Escape Room\\Content\\Audio\\LiveMixingSamples-BassLine.mp3";
			assigned_device_lookup[key].mp3stream = fs.createReadStream(file).pipe(assigned_device_lookup[key].decoder);


			//Function to read file and output streams	 
			var processFrame = function (outputChannels) {	 
				var file_read = new Buffer.allocUnsafe(outputChannels[0].length * 2 * 2);
				console.log("Hello");
				if (assigned_device_lookup[this.key].decoder._readableState.length < file_read.length) {
					//Provide blank data if decoder not ready
					for (let sample = 0; sample < outputChannels[0].length; sample++) {
						outputChannels[0][sample] = buffSamp;
						outputChannels[1][sample] = buffSamp;
					}
					console.log("FILLED BLANKS");
				} else {				
					file_read = assigned_device_lookup[this.key].decoder.read(outputChannels[0].length*2*2);
					console.log(file_read);
					var byte_size = 2;	

					for (let sample = 0; sample < outputChannels[0].length; sample++) {
						outputChannels[0][sample] = file_read.readInt16LE.call(file_read,sample*4);
						outputChannels[1][sample] = file_read.readInt16LE.call(file_read,(sample*4)+2);
					}
					

				}
				return true; // returning false will stop the stream
			}
			//assigned_device_lookup[key].processFrame = func.bind(assigned_device_lookup[key]);
		

			//Audio Worklet
			assigned_device_lookup[key].audioServer = new AudioServer();
			
			assigned_device_lookup[key].outputStream = assigned_device_lookup[key].audioServer.initOutputStream(assigned_device_lookup[key].device.id, {
				format: AudioServer.Int16Array,
				sampleRate: 48000,
				name: "Output_"+assigned_device_lookup[key].device.id,
				//process: assigned_device_lookup[key].processFrame,
				channels: 2
			});
			

			assigned_device_lookup[key].outputStream.start();
			setTimeout(() => {
				console.log('Stopping stream');
				assigned_device_lookup[key].outputStream.stop();
			}, 5000,key);	
		}
	}
}

const main = async () => {
  audioServer.getDevices().outputDevices.forEach(logDevice);
  console.log('-------')

  console.log('API:', audioServer.getApi());
}
main();

```

**Comment by geekuillaume** [comment]
I've changed the API since the version you were using to migrate to a push based method instead of a pull one. I've just pushed a new version (https://github.com/geekuillaume/node-audioworklet/releases/tag/v5.0.3) with the `stream.registerWriteHandler` which mimic the previous `process` option. It should help you migrate your code to this new API.

Once you adapted your code, don't hesitate to tell me if you still have this bug.
